mrmustard/lab_dev/sampler.py

```diff
@@ -0,0 +1,231 @@
+# Copyright 2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Samplers for measurement devices.
+"""
+
+from __future__ import annotations
+from itertools import product
+
+from typing import Sequence, Iterable
+
+import numpy as np
+
+from mrmustard import math, settings
+
+from .states import State, Number
+from .circuit_components import CircuitComponent
+from .circuit_components_utils import BtoQ, TraceOut
+
+__all__ = ["Sampler", "PNRSampler", "HomodyneSampler"]
+
+
+class Sampler:
+    r"""
+    A sampler for measurements of quantum circuits.
+
+    Args:
+        meas_outcomes: The measurement outcomes for this sampler.
+        meas_ops: The measurement operators of this sampler.
+        prob_dist: An optional probability distribution for this sampler.
+    """
+
+    def __init__(
+        self,
+        meas_outcomes: Sequence[any],
+        meas_ops: CircuitComponent | Sequence[CircuitComponent],
+        prob_dist: Sequence[float] | None = None,
+    ):
+        self._meas_ops = meas_ops
+        self._meas_outcomes = meas_outcomes
+        self._prob_dist = prob_dist
```

Comment:
I'm just a little unclear to the role of the `prob_dist` attribute.  I thought it might be used as a cache, but it seems to never be set again after this point.  Shouldn't it be updated when `probabilities()` is called?  Again, I'm just guessing here as to it's role.  Should it only be set on construction?

Reply:
Hmm I dont believe caching the computed probability distribution makes sense as its state dependent so it would only really save us resources if youre repeatedly calling on the same state. 

The idea was if you have a prob dist already you could just set it (e.g for debugging purposes). I think really for this PR I'll remove it in the name of YAGNI and if theres a use case we can reintroduce it